rust_snuba/src/processors/spans.rs

```diff
@@ -0,0 +1,561 @@
+            tags.insert("transaction".into(), transaction.into());
+        }
+
+        if let Some(transaction_op) = &self.transaction_op {
+            tags.insert("transaction.op".into(), transaction_op.into());
+        }
+
+        if let Some(http_method) = &self.http_method {
+            tags.insert("http.method".into(), http_method.into());
+        }
+
+        if let Some(transaction_method) = &self.transaction_method {
+            tags.insert("transaction.method".into(), transaction_method.into());
+        }
+
+        if let Some(status_code) = &self.status_code {
+            tags.insert("status_code".into(), status_code.into());
+        }
+
+        for (key, value) in &self.extra {
+            tags.insert(key.into(), value.into());
+        }
+
+        (
+            tags.keys().cloned().collect(),
+            tags.values().cloned().collect(),
+        )
+    }
+}
+
+#[derive(Debug, Default, Serialize)]
+struct Span {
+    #[serde(default)]
+    action: String,
+    deleted: u8,
+    description: String,
+    #[serde(default)]
+    domain: String,
+    duration: u32,
+    end_ms: u16,
+    end_timestamp: u64,
+    exclusive_time: f64,
+    group: u64,
+    group_raw: u64,
+    is_segment: u8,
+    #[serde(default)]
+    module: String,
+    #[serde(rename(serialize = "measurements.key"))]
+    measurement_keys: Vec<String>,
+    #[serde(rename(serialize = "measurements.value"))]
+    measurement_values: Vec<f64>,
+    offset: u64,
+    op: String,
+    parent_span_id: u64,
+    partition: u16,
+    platform: String,
+    profile_id: Option<Uuid>,
+    project_id: u64,
+    retention_days: u16,
+    segment_id: u64,
+    segment_name: String,
+    #[serde(rename(serialize = "sentry_tags.key"))]
+    sentry_tag_keys: Vec<String>,
+    #[serde(rename(serialize = "sentry_tags.value"))]
+    sentry_tag_values: Vec<String>,
+    span_id: u64,
+    span_kind: String,
+    span_status: u8,
+    start_ms: u16,
+    start_timestamp: u64,
+    status: u32,
+    #[serde(rename(serialize = "tags.key"))]
+    tag_keys: Vec<String>,
+    #[serde(rename(serialize = "tags.value"))]
+    tag_values: Vec<String>,
+    trace_id: Uuid,
+    transaction_id: Uuid,
+    transaction_op: String,
+    user: String,
+}
+
+impl TryFrom<FromSpanMessage> for Span {
+    type Error = BadMessage;
+
+    fn try_from(from: FromSpanMessage) -> Result<Span, BadMessage> {
+        let end_timestamp_ms = from.start_timestamp_ms + from.duration_ms as u64;
+        let group = if let Some(group) = &from.sentry_tags.group {
+            u64::from_str_radix(group, 16).map_err(|_| BadMessage)?
+        } else {
+            0
+        };
+        let status = from.sentry_tags.status.unwrap_or_default() as u8;
+        let (sentry_tag_keys, sentry_tag_values) = from.sentry_tags.to_keys_values();
+        let transaction_op = from.sentry_tags.transaction_op.unwrap_or_default();
+
+        let tags = from.tags.unwrap_or_default();
```

Comment:
in that case you can also have serde `default` that, instead of making it an `Option<BTreeMap>`